### ISSUE TITLE
[rllib] How do you actually use variable sized action spaces with (e.g.) PPOAgent

### ISSUE THREAD

**Issue by henrycharlesworth** [ray-project/ray#4089]
I have a custom multi-agent Poker environment I'm working on where the number of actions available depends on the state. I tried to mimic the <a href="https://github.com/ray-project/ray/blob/master/python/ray/rllib/examples/parametric_action_cartpole.py">parametric cartpole</a> example but it's not really clear to me how this is working.

I've set up my environment to return observations in the form:
`obs = {"player": {"actions_mask": ..., "real_obs": ...}, "opponent": {"actions_mask": ..., "real_obs": ...}`
with "real_obs" corresponding to the actual neural network input and "actions_mask" being a vector of 0s (action unavailable) and 1s (action available).

I then make a custom model as follows (very similar to the parametric_cartpole example):
`class pokerModel(Model):

    def _build_layers_v2(self, input_dict, num_outputs, options):
        actions_mask = input_dict["obs"]["actions_mask"]

        #standard FC net:
        last_layer = input_dict["obs"]["real_obs"]
        hiddens = [512, 256]
        for i, size in enumerate(hiddens):
            label = "fc{}".format(i)
            last_layer = slim.fully_connected(last_layer, size,
                                              weights_initializer=normc_initializer(1.0), activation_fn=tf.nn.relu, scope=label)
        output = slim.fully_connected(last_layer, num_outputs, activation_fn=None, scope="fc_out")

        inf_mask = tf.maximum(tf.log(actions_mask), tf.float32.min)
        masked_logits = inf_mask + output

        return masked_logits, last_layer`

My question is can I (and if so how) use this model with the PPOPolicyGraph? Or how do I use it within a PPOAgent? I register the model with the name "poker_model" and then try doing something like this to try generating some batches:

`from ray.rllib.evaluation.policy_evaluator import PolicyEvaluator

ev = PolicyEvaluator(env_creator=lambda _: HeadsUpGame(),
       policy_graph={
                "player": (PPOPolicyGraph, obs_space, act_space, {"model":{
                        "custom_model":"poker_model"
                   }}),
                  "opponent": (PPOPolicyGraph, obs_space, act_space, {"model":{
                          "custom_model":"poker_model"
                    }})
                     },
                     policy_mapping_fn=lambda agent_id: "player",
                    batch_steps=10) 
batch = ev.sample()`

but it doesn't work. When I look at the PPOPolicyGraph source code, to load the model it does the following:

`self.model = ModelCatalog.get_model(
            {
                "obs": obs_ph,
                "prev_actions": prev_actions_ph,
                "prev_rewards": prev_rewards_ph,
                "is_training": self._get_is_training_placeholder(),
            },
            observation_space,
            logit_dim,
            self.config["model"],
            state_in=existing_state_in,
            seq_lens=existing_seq_lens)`

where `obs_ph = tf.placeholder(
                tf.float32,
                name="obs",
                shape=(None, ) + observation_space.shape)`
is defined in the init function. But to actually load my custom model in a way that works I need to do:
`p1 = ModelCatalog.get_model({
        "obs": {
            "real_obs": tf.placeholder(tf.float32, shape=(None, 308)),
            "actions_mask": tf.placeholder(tf.float32, shape=(None, 23))
        }
    }, Box(0, 1, shape=(308,), dtype=np.float32), 23, {"custom_model": "poker_model"})`

Without writing a new PPOPolicyGraph class what should I do to use my custom model with the PPO algorithm? Clearly this is possible as the parametric_cartpole example runs, but it's not clear how the run_experiments function is generating the PPOAgent.

**Comment by RylanSchaeffer** [comment]
@henrycharlesworth why did you close this? If you found the solution, can you please share it?

**Comment by henrycharlesworth** [comment]
@RylanSchaeffer I closed it right after I submitted it because I realised I should have posted it in the ray-dev google group. I asked it here instead and got an answer: https://groups.google.com/forum/?nomobile=true#!topic/ray-dev/_cBxgfHTYZw
